**DATA INGESTION**

In [ ]:
CREATE TABLE DEMO.dbo.Titanic(
	PassengerId INT IDENTITY(1,1) primary key,
	Survived INT,
	Pclass INT,
	Name NVARCHAR(200),
	Sex NVARCHAR(10),
	Age FLOAT,
	SibSp INT,
	Parch INT,
	Ticket NVARCHAR(50),
	Fare FLOAT,
	Cabin NVARCHAR(50),
	Embarked NVARCHAR(50))
GO

In [ ]:
ALTER TABLE DEMO.dbo.Titanic


/*Training Dataset*/

BULK INSERT DEMO.dbo.Titanic
FROM '.\titanic_dataset\train.csv'
WITH
(
        FORMAT='CSV',
        FIRSTROW=2
)
GO

In [28]:
USE [DEMO]

Commands completed successfully.

Total execution time: 00:00:00

In [17]:
-- view TOP 20 of the data
SELECT TOP(20) * FROM Titanic 

(20 rows affected)

Total execution time: 00:00:00.018

PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked 1 0 3 Braund, Mr. Owen Harris male 22 1 0 A/5 21171 7.25 NULL S 2 1 1 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38 1 0 PC 17599 71.2833 C85 C 3 1 3 Heikkinen, Miss. Laina female 26 0 0 STON/O2. 3101282 7.925 NULL S 4 1 1 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35 1 0 113803 53.1 C123 S 5 0 3 Allen, Mr. William Henry male 35 0 0 373450 8.05 NULL S 6 0 3 Moran, Mr. James male NULL 0 0 330877 8.4583 NULL Q 7 0 1 McCarthy, Mr. Timothy J male 54 0 0 17463 51.8625 E46 S 8 0 3 Palsson, Master. Gosta Leonard male 2 3 1 349909 21.075 NULL S 9 1 3 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27 0 2 347742 11.1333 NULL S 10 1 2 Nasser, Mrs. Nicholas (Adele Achem) female 14 1 0 237736 30.0708 NULL C 11 1 3 Sandstrom, Miss. Marguerite Rut female 4 1 1 PP 9549 16.7 G6 S 12 1 1 Bonnell, Miss. Elizabeth female 58 0 0 113783 26.55 C103 S 13 0 3 Saundercock, Mr. William Henry male 20 0 0 A/5. 2151 8.05 NULL S 14 0 3 Andersson, Mr. Anders Johan male 39 1 5 347082 31.275 NULL S 15 0 3 Vestrom, Miss. Hulda Amanda Adolfina female 14 0 0 350406 7.8542 NULL S 16 1 2 Hewlett, Mrs. (Mary D Kingcome) female 55 0 0 248706 16 NULL S 17 0 3 Rice, Master. Eugene male 2 4 1 382652 29.125 NULL Q 18 1 2 Williams, Mr. Charles Eugene male NULL 0 0 244373 13 NULL S 19 0 3 Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele) female 31 1 0 345763 18 NULL S 20 1 3 Masselmani, Mrs. Fatima female NULL 0 0 2649 7.225 NULL C

In [21]:
-- How many samples are present in the training data
SELECT
COUNT(*)
    total_num_of_samples
FROM 
    Titanic ;

(1 row affected)

Total execution time: 00:00:00.007

total_num_of_samples 891

In [41]:
-- check for null values in the target column

SELECT 
    COUNT(*)
FROM
    Titanic
WHERE
Survived IS NULL

(1 row affected)

Total execution time: 00:00:00.115

(No column name) 0

In [57]:
USE [Demo];
GO

CREATE PROCEDURE CheckNulls
@column nvarchar(10)
AS
SELECT COUNT(*) FROM Titanic WHERE @column IS NULL

Commands completed successfully.

Commands completed successfully.

Total execution time: 00:00:00.033

In [67]:
EXEC CheckNulls @column = 'Pclass';

(1 row affected)

Total execution time: 00:00:00.013

(No column name) 0

In [39]:
SELECT TOP(1)
    (SELECT COUNT(*) FROM Titanic WHERE Pclass IS NULL) Pclass,
    (SELECT COUNT(*) FROM Titanic WHERE Sex IS NULL) Sex,
    (SELECT COUNT(*) FROM Titanic WHERE Age IS NULL) Age,
    (SELECT COUNT(*) FROM Titanic WHERE SibSP IS NULL) SibSP,
    (SELECT COUNT(*) FROM Titanic WHERE Parch IS NULL) Parch,
    (SELECT COUNT(*) FROM Titanic WHERE Ticket IS NULL) Ticket,
    (SELECT COUNT(*) FROM Titanic WHERE Name IS NULL) Name,
    (SELECT COUNT(*) FROM Titanic WHERE Fare IS NULL) Fare,
    (SELECT COUNT(*) FROM Titanic WHERE Cabin IS NULL) Cabin,
    (SELECT COUNT(*) FROM Titanic WHERE Embarked IS NULL) Embarked
FROM
    Titanic

(1 row affected)

Total execution time: 00:00:00.402

Pclass Sex Age SibSP Parch Ticket Name Fare Cabin Embarked 0 0 177 0 0 0 0 0 687 2

In [42]:
USE [DEMO]
SELECT
    COUNT(*) counts
FROM
    TitaNIC
GROUP BY
    Survived

--OR, using correlated sub-queries (works well for Azure Data Studio Visualization)

SELECT TOP(1)
    (SELECT COUNT(*) FROM Titanic WHERE Survived = 0) AS Num_of_recorded_deaths,
    (SELECT COUNT(*) FROM Titanic WHERE Survived = 1) AS Num_of_survived_passengers
FROM
    Titanic

(2 rows affected)

(1 row affected)

Total execution time: 00:00:00.019

counts 549 342

Num_of_recorded_deaths Num_of_survived_passengers 549 342

In [109]:
-- summary statistics
CREATE PROCEDURE describe

AS

SELECT
    AVG(AGE) avg_age,
    MAX(Sex) gender_most_frequent,
    MAX(Pclass) pclass_most_frequent,
    MAX(Embarked) embarked_most_frequent,
    MAX(Cabin) cabin_most_frequent
FROM Titanic;

Commands completed successfully.

Total execution time: 00:00:00.083

In [108]:
SELECT * FROM sys.objects WHERE Object_id = OBJECT_ID('demo.dbo.Titanic')

(1 row affected)

Total execution time: 00:00:00.025

name object_id principal_id schema_id parent_object_id type type_desc create_date modify_date is_ms_shipped is_published is_schema_published Titanic 965578478 NULL 1 0 U USER_TABLE 2022-10-19 09:34:13.807 2022-10-19 09:34:13.807 0 0 0

In [111]:
EXEC describe;

Warning: Null value is eliminated by an aggregate or other SET operation.

(1 row affected)

Total execution time: 00:00:00.080

avg_age gender_most_frequent pclass_most_frequent embarked_most_frequent cabin_most_frequent 29.69911764705882 male 3 S T

Report

```
There are no missing values in the target column (Survived)
```
```
There are a total of 177 and 687 missing values in the Age column and Cabin column respectively.

```

**DATA CLEANING**

In [6]:
-- create a temporary #Titanic_temp for data cleaning...

SELECT *
INTO 
    #Titanic_temp
FROM
    Demo.dbo.Titanic

(891 rows affected)

Total execution time: 00:00:00.784

In [8]:
CREATE PROCEDURE head
@n int
AS
SELECT TOP(@n) * FROM #Titanic_temp;

Commands completed successfully.

Total execution time: 00:00:00.226

In [9]:
EXECUTE head @n = 5;

(5 rows affected)

Total execution time: 00:00:00.013

PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked 1 0 3 Braund, Mr. Owen Harris male 22 1 0 A/5 21171 7.25 NULL S 2 1 1 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38 1 0 PC 17599 71.2833 C85 C 3 1 3 Heikkinen, Miss. Laina female 26 0 0 STON/O2. 3101282 7.925 NULL S 4 1 1 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35 1 0 113803 53.1 C123 S 5 0 3 Allen, Mr. William Henry male 35 0 0 373450 8.05 NULL S

In [37]:
--Explore the Percentage of missing values in Age and Cabin and Embarked 

Use [Demo];

SELECT TOP(1)
    (
        CAST(
            (SELECT COUNT(*) FROM #Titanic_temp WHERE Cabin is NULL) AS FLOAT)/
            COUNT(*))*100 as '% missing_cabin',
    (
        CAST(
            (SELECT COUNT(*) FROM #Titanic_temp WHERE Age is NULL) AS FLOAT)/
            COUNT(*))*100 as '% missing_age',

    (
        CAST(
            (SELECT COUNT(*) FROM #Titanic_temp WHERE Embarked is NULL) AS FLOAT)/
            COUNT(*))*100 as '% missing_embarked'

    FROM #Titanic_temp

(1 row affected)

Total execution time: 00:00:00.024

% missing_cabin % missing_age % missing_embarked 77.10437710437711 19.865319865319865 0.22446689113355783

> Cabin contains 77% of missing values while Age and Embarked contain 19.9% and 0.002244 of missing values respectively

<b\> Imputation Steps</b>
 
<ul>

    <li>Cabin contains too missing values and should be droped to avoid complexity and wrong imputations </li>

    <li\>Fill Null's in Embarked with the modal category </li>
    <li\>Fill Null's in Age with average age of each class in Survived column</li>
</ul\>

In [82]:
IF (
    SELECT 
    COUNT(*)
    FROM
        #Titanic_temp
    WHERE
        Embarked IS NULL)
<> 0
BEGIN
    UPDATE #Titanic_temp
    SET Embarked = (SELECT MAX(Embarked) FROM #Titanic_temp)
    WHERE Embarked IS NULL
    PRINT 'Imputing missing values in Embarked column with its mode'
END

ELSE 
BEGIN
    PRINT('imputing already done')
END 


Warning: Null value is eliminated by an aggregate or other SET operation.

(2 rows affected)

Imputing missing values in Embarked column with its mode

Total execution time: 00:00:00.159

In [83]:
-- Check operation's success
SELECT COUNT(*) FROM #Titanic_temp WHERE Embarked IS NULL

(1 row affected)

Total execution time: 00:00:00.008

(No column name) 0

**EXPLORATORY DATA ANALYSIS**

In [19]:
-- a closer look at passengers origin
SELECT
    Embarked, COUNT(*) counts 
FROM 
    Titanic
WHERE
    Embarked is NOT NULL
GROUP BY
    Embarked

(3 rows affected)

Total execution time: 00:00:00.050

Embarked counts S 644 Q 77 C 168

In [24]:
-- a closer look at passengers origin
SELECT
    Embarked, COUNT(*) counts 
FROM 
    Titanic
WHERE
    Embarked is NOT NULL
GROUP BY
    Embarked

(3 rows affected)

Total execution time: 00:00:00.008

Embarked counts S 644 Q 77 C 168

In [25]:
SELECT
    SibSp, COUNT(*) counts 
FROM 
    Titanic
WHERE
    SibSP is NOT NULL
GROUP BY
    SibSp
ORDER BY
    SibSp;

(7 rows affected)

Total execution time: 00:00:00.018

SibSp counts 0 608 1 209 2 28 3 16 4 18 5 5 8 7

In [26]:
SELECT
    Parch, COUNT(*) counts 
FROM 
    Titanic
WHERE
    Parch is NOT NULL
GROUP BY
    Parch
ORDER BY
    Parch;

(7 rows affected)

Total execution time: 00:00:00.019

Parch counts 0 678 1 118 2 80 3 5 4 4 5 5 6 1

In [5]:
SELECT TOP(10) * FROM [Demo].dbo.Titanic

(10 rows affected)

Total execution time: 00:00:00.213

PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked 7 0 1 McCarthy, Mr. Timothy J male 54 0 0 17463 51.8625 E46 S 8 0 3 Palsson, Master. Gosta Leonard male 2 3 1 349909 21.075 NULL S 9 1 3 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27 0 2 347742 11.1333 NULL S 10 1 2 Nasser, Mrs. Nicholas (Adele Achem) female 14 1 0 237736 30.0708 NULL C 1 0 3 Braund, Mr. Owen Harris male 22 1 0 A/5 21171 7.25 NULL S 2 1 1 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38 1 0 PC 17599 71.2833 C85 C 3 1 3 Heikkinen, Miss. Laina female 26 0 0 STON/O2. 3101282 7.925 NULL S 4 1 1 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35 1 0 113803 53.1 C123 S 5 0 3 Allen, Mr. William Henry male 35 0 0 373450 8.05 NULL S 6 0 3 Moran, Mr. James male NULL 0 0 330877 8.4583 NULL Q

In [15]:
SELECT
    Survived,
    Sex,
    COUNT(Sex) count_sex,
    AVG(Cast(Age as int)) avg_age
FROM
    Demo.dbo.Titanic
WHERE Sex IS NOT NULL
AND
Survived is NOT NULL
GROUP BY
    Survived,
    Sex
ORDER BY SEX


Warning: Null value is eliminated by an aggregate or other SET operation.

(4 rows affected)

Total execution time: 00:00:00.037

Survived Sex count_sex avg_age 0 female 81 25 1 female 233 28 0 male 468 31 1 male 109 27